File: 25 Feb 2024/sort_0_1_2.py
def sort012(arr, n):
    low, mid, high = 0, 0, n -1

    while mid <= high:

        if arr[mid] == 0:
            arr[mid], arr[low] = arr[low], arr[mid]
            mid += 1
            low += 1
        
        elif arr[mid] == 1:
            mid += 1
        
        else:
            arr[high], arr[mid] = arr[mid], arr[high]
            high -= 1

a = [0, 2, 1, 2, 0]

sort012(a, len(a))

print(a)

File: 3 April 2024/sort_0_1_2.cpp
// sort and array of 0, 1 and 2

#include <bits/stdc++.h>
using namespace std;

int main(int argc, char *argv[])
{
    int n = 0;
    vector<int> arr;
    for (int i = 1; i < argc; i++)
    {
        arr.push_back(atoi(argv[i]));
        n++;
    }

    for (int e : arr)
        cout << e << " ";

    cout << endl
         << n << endl;

    int temp, low = 0, mid = 0, high = n - 1;

    while (mid <= high)
    {
        switch (arr[mid])
        {
        case 0:
            temp = arr[low];
            arr[low] = arr[mid];
            arr[mid] = temp;
            mid++;
            low++;
            break;
        case 1:
            mid++;
            break;
        case 2:
            temp = arr[high];
            arr[high] = arr[mid];
            arr[mid] = temp;
            mid++;
            high--;
            break;
        default:
            break;
        }
    }

    for (int e : arr)
        cout << e << " ";

    return 0;
}